### ISSUE TITLE
How to distribute various z/OS and VM/CMS builds

### ISSUE THREAD

**Issue by pmqs** [pmqs/z1#268]
### Posted by Al Dunsmuir » Sat Jul 24, 2010 5:15 pm
Folks,\
One of the unresolved areas of bringing the z/OS and VM/CMS ports up to speed is how to make these packages (especially pre-built binaries) available for testing (functional and performance).

I checked the [www.infozip.org](http://www.infozip.org) web page, and there are currently no binaries available from there for z/OS MVS, z/OS USS or VM/CMS.  The ftp sites linked from those ftp sites appear to be down for all platforms.  The z/OS USS version is apparently available elsewhere (old release from an IBM web page, current from ?).  The latest VM/CMS version available for download elsewhere is apparently from 1998.

It seems like what I'm experiencing is a general problem, and so requires a consistent general solution that would best be decided by Ed, SMS and Hunter.

As to the mainframe ports, it seems to me to be necessary to have the current working code for these platforms available as a baseline.  This would include the last 32-bit zip and unzip releases, and the current 64-bit zip and unzip releases.      Having this would also provide some measure of relief for z/OS MVS and VM/CMS users until we have well-tested ports of the latest zip/unzip beta available with my changes integrated.

I'm proposing that I should do z/OS MVS, z/OS USS and VM/CMS builds for the following:

* the last 32-bit releases of zip 2.32/unzip
* that release with my z/OS MVS deltas.
* the latest zip/unzip betas with 64-bit support (if possible with only minor fixes)

Those builds would be used to establish a baseline before starting integrating the changes I currently have available into the latest beta zip and unzip codebases.

Managing comminications or links to these baseline and iterative test releases for the various ports could get rather confusing at worst, or messy at best.  While there will need to be test builds for each of the Ed and SMS's betas there will be many more test builds as new port-specific function is developed, unit tested, and validated.

My suggestions to Ed, SMS and Hunter would be to create one of the following:

* Within the **InfoZIP Announcements** category, between **InfoZIP Releases** and Site Discussion add a new InfoZIP Test Releases forum, where I can create a topic for zip and another for unzip per port.

* Below the **InfoZIP Bugs** category, create a new **InfoZIP Test Releases** category, with the 3 standard forums (**InfoZIP Zip**, **InfoZIP UnZip** and **InfoZIP WIZ**.    In each forum, I could a create a topic for that tool for a given port.

This is the one I would prefer, as it mirrors the current separation of Zip and UnZip discussions.

Either of these setups would help keep things sane.

It seems important that the new test release forums would have a small but prominate disclaimer that test releases are for testing only, and should not be redistributed, or used for production work.

Once the organization of discussion is set up, the next question is how the actual port should be available for download.  Should these be an attachment within the forum topic, or should the topic refernce files available via FTP from SourceForge.  The latter seems better, since we can make files with known problems "go away". In that case, I would need access to create those files.

Al

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Sat Jul 24, 2010 5:30 pm
I saw a posting that identified <ftp://ftp.info-zip.org/pub/infozip/> as the download site.

* A check shows there are indeed some packages under the mvs and vmcms directories.

The web page at [www.infozip.org](http://www.infozip.org) should probably be updated to replace the 3 dead sites with the real ftp site name.

**Comment by pmqs** [comment]
### Posted by E G » Mon Jul 26, 2010 7:16 pm
Should this area be for official Info-ZIP postings or for unofficial postings that anyone can post?  I'm thinking the latter.  Currently the only official places to get Info-ZIP stuff are the ftp site (noted above) and the Info-ZIP area on SourceForge.  It's possible that postings here, once tested, could then work their way to the official places.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Mon Jul 26, 2010 9:07 pm
>**E G** wrote:\
>Should this area be for official Info-ZIP postings or for unofficial postings that anyone can post?  I'm thinking the latter.


Maximizing the number of actively maintained ports is one of the goals of making platform-specific builds available for test (and later distribution).
If we have restrictions on who can post to the forum, we limit the ability for folks to participate... and for some platforms suitable developers may be few and far between.

My preferred suggestion was to ask Hunter to create a new InfoZIP Test Releases category with the 3 standard forums (**InfoZIP Zip**, **InfoZIP UnZip** and **InfoZIP WIZ**) below the existing **InfoZIP Bugs** category.

Within  each  Forum, we would want one  topic  per  platform,  with the    first    posting  in the topic conforming to a standard layout with:

* Platform short name
* Platform description, including relevant URLs.
* The relevant  URL(s) in the InfoZIP FTP site for tested builds for a given release (or beta)
* The relevant URL(s) in the InfoZIP FTP site, other FTP site, or web page for test builds.
* A standard disclaimer about using the test builds  for  production purposes.

We would  end  up  with a topic  each for 32-bit Linux, 64-bit Linux, Mac OS/X,  32-bit  Windows, 64-bit Windows, z/OS MVS, z/OS USS, VM/CMS,  Windows CE, Solaris, and all the others.

Each topic would be created the first time someone decides to pony-up and  supply the first build for a given platform.  It would be useful but not essential if Hunter could set up the Forum software could supply a standardized template with the standard information and disclaimer all ready to customize when a new topic is created (and the topic subject assigned) under this category.

Subsequent new builds for that platform and the promotion of builds to "tested' and available from the InfoZIP web site would all be announced by additional posting within the existing topic.


>**E G** wrote:\
>Currently the only official places to get Info-ZIP stuff are the ftp site (noted above) and the Info-ZIP area on SourceForge.  It's possible that postings here, once tested, could then work their way to the official places.


Promoting builds that have already been tested through a clearly defined process sounds ideal.

Some folks doing platform builds may already have a public ftp site which they can use to distribute test builds.    When a build is dubbed "tested", it needs to be promoted to the InfoZIP web site.    The builder would either need access to that platform's subdirectories on the InfoZIP web site, or would get someone on the InfoZIP team to do the promotion.

Other platform builders (such as myself) would not have access to a public ftp site.  In that case, what has been done in the past is to grant authorizatio to a "beta" subdirectory on the InfoZIP web site under their platform.    Promotions would be by the builder (if authorized) or the InfoZIP team.

One advantage of keeping announcements for new builds central is that helps minimize the risk of some varmit distributing an evil variant of zip/or unzip  from their own web site.    If something bad does happen (such as the attack on Berlios), it allows the InfoZIP team to quickly shut down links to the compromized site and calmly describe the problem and  procedure folks should folllow (revert to previous level, or install corrected build)

**Comment by pmqs** [comment]
### Posted by E G » Mon Jul 26, 2010 10:03 pm
>**Al Dunsmuir** wrote:\
>My preferred suggestion was to ask Hunter to create a new InfoZIP Test Releases category with the 3 standard forums (**InfoZIP Zip**, **InfoZIP UnZip** and **InfoZIP WIZ**) below the existing **InfoZIP Bugs** category.
>
> Within  each  Forum, we would want one  topic  per  platform,  with the    first    posting  in the topic conforming to a standard layout with:
>
>   * Platform short name
>   * Platform description, including relevant URLs.
>   * The relevant  URL(s) in the InfoZIP FTP site for tested builds for a given release (or beta)
>   * The relevant URL(s) in the InfoZIP FTP site, other FTP site, or web page for test builds.
>   * A standard disclaimer about using the test builds  for  production purposes.
>
We would  end  up  with a topic  each for 32-bit Linux, 64-bit Linux, Mac OS/X,  32-bit  Windows, 64-bit Windows, z/OS MVS, z/OS USS, VM/CMS,  Windows CE, Solaris, and all the others.


If no one has comments to the contrary, sounds like a plan.  Hunter is definitely the one to set it up.  When you're ready, just ask him.



>**Al Dunsmuir** wrote:\
>Each topic would be created the first time someone decides to pony-up and  supply the first build for a given platform.  It would be useful but not essential if Hunter could set up the Forum software could supply a standardized template with the standard information and disclaimer all ready to customize when a new topic is created (and the topic subject assigned) under this category.


It might be better to prevent outsiders from creating new threads in this new forum test area then.  We can add an additional thread for platform addition requests and any other maintenance users want done.  That way there will be no additional threads with random topics popping up in the test areas.  We have the ability to rename, move, and delete threads, though that really hasn't been done.

So Info-ZIP would create each new platform thread and make sure it follows the rules.  It would be good if Hunter can set it up so that Info-ZIP forum members can create threads in the test area.



>**Al Dunsmuir** wrote:\
>Subsequent new builds for that platform and the promotion of builds to "tested' and available from the InfoZIP web site would all be announced by additional posting within the existing topic.


Probably also can announce in the Release area as well.  That area probably should be cleaned up some once this is set up.



>**Al Dunsmuir** wrote:\
>Promoting builds that have already been tested through a clearly defined process sounds ideal.


We'll have to come up with some criteria for when a platform is sufficiently tested and stable.

>**Al Dunsmuir** wrote:\
> Some folks doing platform builds may already have a public ftp site which they can use to distribute test builds.    When a build is dubbed "tested", it needs to be promoted to the InfoZIP web site.    The builder would either need access to that platform's subdirectories on the InfoZIP web site, or would get someone on the InfoZIP team to do the promotion.


Currently only a few have the ability to put stuff on the ftp and SourceForge sites.  Probably best to keep it that way.  Should be no problem copying a build from the forum to those locations.  Currently any official postings are vetted by the development group before they are posted.  In cases where no one in the group has access to a compiler or platform to test the build, we should note that we haven't verified the build.  Maybe by adding a note in the build distribution.  In fact, maybe it makes sense to always add a BUILDNOTES.TXT file, or something like that, to every platform distribution with the details of where it came from and what testing was done.

We'll also have to decide what goes in a distribution.  Currently a platform distribution, such as for Windows, is an archive with the documentation in the correct line end for the platform, any executables (such as zip, zipnote, zipcloak, and zipsplit for Zip distributions), and any additional information relevant for that platform.



>**Al Dunsmuir** wrote:\
>Other platform builders (such as myself) would not have access to a public ftp site.  In that case, what has been done in the past is to grant authorizatio to a "beta" subdirectory on the InfoZIP web site under their platform.    Promotions would be by the builder (if authorized) or the InfoZIP team.


By Info-ZIP web site, what do you mean?  This forum is part of the web site.  The FTP site as well.  SourceForge is separate, but more or less everything posted there comes from the FTP site where things are posted first.



>**Al Dunsmuir** wrote:\
>One advantage of keeping announcements for new builds central is that helps minimize the risk of some varmit distributing an evil variant of zip/or unzip  from their own web site.    If something bad does happen (such as the attack on Berlios), it allows the InfoZIP team to quickly shut down links to the compromized site and calmly describe the problem and  procedure folks should folllow (revert to previous level, or install corrected build)


Hopefully build distributions will be vetted well before they get put on the official site.  Anything in the forum should be considered experimental.  Once vetted, a link can be added to the forum thread pointing to the official distribution for that platform on the FTP site.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Mon Jul 26, 2010 11:50 pm
>**E G** wrote:\
>If no one has comments to the contrary, sounds like a plan.  Hunter is definitely the one to set it up.  When you're ready, just ask him.


Ed,

It appears that the structure is there, but we need to zero in on how to manage access.  There seems to be a bit of indecision yet regarding making the new category wide open to forum members, or restricted.    I suggest that we have Hunter set it up with "InfoZIP Team"-only able to post and create new topics.  We can expand that as things coalesce.

We need to get comments from folks interested in being platform builders.  If we set it up the right way, it will encourage folks to become platform builders, and perhaps later become part of the core team.  If we set it up with too much overhead, it won't accomplish this goal.  I think this is critical to getting InfoZIP more visibility, credance and assistance in these days where there are many competing compression and archive tools.



>**E G** wrote:\
>It might be better to prevent outsiders from creating new threads in this new forum test area then.  We can add an additional thread for platform addition requests and any other maintenance users want done.  That way there will be no additional threads with random topics popping up in the test areas.  We have the ability to rename, move, and delete threads, though that really hasn't been done.
>
> So Info-ZIP would create each new platform thread and make sure
>  it follows the rules.  It would be good if Hunter can set it up so that
>  Info-ZIP forum members can create threads in the test area.


Currently there are three categories of users on this forum - general, "InfoZIP Team" (to which I thank yourself and Hunter for promoting me) and "Administrator Group" (Hunter).    Anyone in the later two groups should be able to create new topics in the Forums under the new "InfoZIP Test Releases" category.

I propose the following:


* A special topic "New InfoZIP Port Requests" be created in each forum under the new "InfoZIP Test Releases" category, which would be made postable by everyone.  This would be where folks would go to define a new (or existing) post for which no topic exists.

  I don't know if Hunter can do this for one topic, while restricting who can create or post to other topics.


* A fourth team defined - "InfoZIP Build Team".  This group would be able to post to an existing topic  (existing port) under the new "InfoZIP Test Releases" category.

  They should also be able to post in any of the topics in this category, providing information about new or updated builds.

  I would suggest that folks in this group be able to add new topics (representing a new port).  They would be required to follow the rules for documenting and describing their port.   Folks in the "InfoZIP Team" can revise/revert these if required.

  This makes it very convenient for a known platform builder to be able to add additional ports for initial user test without assistance.



>**E G** wrote:\
>Probably also can announce in the Release area as well.  That area probably should be cleaned up some once this is set up.



I don't know if the "InfoZIP Team" has the resources to vett each release personally.  They should certainly be involved with promoting a test release to the production area (old production releases never go away), and ensuring it is complete and follows the format agreed for that platform.



>**E G** wrote:\
>We'll have to come up with some criteria for when a platform is sufficiently tested and stable.

Yes.

Verification that the port has correctly processed some common test zip files.
Verification that the port supports all platform-specific extensions.
Any known bugs and limitations should also be documented.



>**E G** wrote:\
>Currently only a few have the ability to put stuff on the ftp and SourceForge sites.  Probably best to keep it that way.


Sounds reasonable for the production releases, however as I've noted the z/OS and CMS releases have historically used a subdirectory under the main directory on the ftp site to distribute platform-specific files (test builds of beta code).

I propose that Hunter set up ACLs for the ftp directories that would allow the main folks to update anywhere in the ftp site, and to add specific folks as R/W to the required port-specific directories.   This way the betas can be efficiently distributed (not via attachments) and promotion is controlled.



>**E G** wrote:\
>Should be no problem copying a build from the forum to those locations.  Currently any official postings are vetted by the development group before they are posted.  In cases where no one in the group has access to a compiler or platform to test the build, we should note that we haven't verified the build.  Maybe by adding a note in the build distribution.


It should be clearly document who has verified the build (EG, specific members of the "InfoZIP Build team"), and who has not.



>**E G** wrote:\
>In fact, maybe it makes sense to always add a BUILDNOTES.TXT file, or something like that, to every platform distribution with the details of where it came from and what testing was done.


Yes.  Pointers to the forum would also serve well in this role.  In that case, BUILDNOTES.HTML might be better (simple HTML, done via editor).



>**E G** wrote:\
>We'll also have to decide what goes in a distribution.  Currently a platform distribution, such as for Windows, is an archive with the documentation in the correct line end for the platform, any executables (such as zip, zipnote, zipcloak, and zipsplit for Zip distributions), and any additional information relevant for that platform.



Indeed.


>**E G** wrote:\
>By Info-ZIP web site, what do you mean?  This forum is part of the web site.  The FTP site as well.  SourceForge is separate, but more or less everything posted there comes from the FTP site where things are posted first.


The forums are controlled by one set of software, with it's own rules and access controls.  The ftp site and web site are controlled by the ftp server and webserver respectively, with their own separate access controls. While it may logically form a web site, each to me is a separate component of that site.



>**E G** wrote:\
>Hopefully build distributions will be vetted well before they get put on the official site.  Anything in the forum should be considered experimental.  Once vetted, a link can be added to the forum thread pointing to the official distribution for that platform on the FTP site.




For  platforms where the users are less able to do builds (whether due to  the  build  setup  being  difficult/obscure  or  due  to the build software  being  proprietary  and  costing  $$s),  making the platform builds easier to get will help keep zip/unzip users involved.


For  the  other  platforms, having "reference builds" are helpful as a baseline  for  bug  reports,  and  to demonstrate that the platform is actively  supported.  For  cases  like Linux, you can likely get folks involved with the distributions to help in this task. The same for the folks involved with the GNUWin32 project on 32-bit windows.

I think that we've got a reasonably good handle on builds and builders if we can come to agreement on the above.




It makes to restrict the new "InfoZIP Test Releases" category posts strictly controlled and dedicated to "InfoZIP Team" and "InfoZIP Build Team" announcements.

We come back to how to we make it easy for users to post regarding validation of a given platform build (good news, or bad).  We need lots of active user testing, with posts somewhere that validate  given builds along the lines of  "it works", "it fails".  It makes sense to keep these posts separate from the formal release announcements, test release announcements, and design discussion - if we succeed in getting widespread testing and input from users on various platforms, it may be high volume, so best to plan ahead.

It appears that we may have yet another reason to add a new category.  We could have:

* InfoZIP Announcement
* InfoZIP Software Discussion and Feature Requests
* InfoZIP Bugs
* InfoZIP Test Releases
* InfoZIP Test Release Discussion

In the latter category, we would have the traditional 3 Forums.  Anyone in the "InfoZIP Team" and "InfoZIP Build Team" would be able to create a new topic.  It would be best if these topics matched those in the "InfoZIP Test Release" forum.

This gives a place for users within a given platform to identify their success or lack thereof when using that platform's test releases.  It would provide a place to have an initial discussion about platform-specific bugs or even common bugs in a given build for a given port.  Detailed discussion of new features or bugs (whether common or platform specific) would move over to the existing forum categories, as is done now.

What does everyone (and especially Ed, Steven and Hunter) think?

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Tue Jul 27, 2010 12:11 am
I just had another idea for distributing builds via the [ftp.info-zip.org](ftp://ftp.info-zip.org) site.

The current directory structure for each platform supported by InfoZIP is:

platform_name

* That directory contains the current and old formal releases
* Some platforms have a "beta" directory under that.

I would suggest that this be replaced by the following:

platform_name

* release        - Contains current and old releases.
* beta           - Contains the tested beta release for a given official beta
* test           - Contains test releases for new beta and platform code

The promotion path would be test->beta->release with no ambiguity.

Al

**Comment by pmqs** [comment]
### Posted by E G » Tue Jul 27, 2010 2:47 am
OK, I can go back and forth a little more, but I got to get back to beta development shortly.

The format I'm using below is:

Category
* Forum - forum description
  * Forum Subject (also called thread or topic)
    * Postings in the thread (comments)


Here are the existing categories available to users:

* Info-ZIP Announcements
  * Info-ZIP Releases        - Info-ZIP release announcements
  * Site Discussions          - Problems or comments about <http://www.info-zip.org>

* Info-ZIP Software Discussions and Feature Requests
  * Info-ZIP Zip                  - Discussions and questions about Info-ZIP Zip.
  * Info-ZIP UnZip              - Discussions about Info-ZIP UnZip.
  * Info-ZIP WiZ                  - Discussions about Info-ZIP WiZ.

* Info-ZIP Bugs
  * Zip Bugs                          - Please report Zip bugs here!
  * UnZip Bugs                      - Please report UnZip bugs here.
  * WiZ Bugs                          - Please report WiZ bugs here.

Under each forum are subjects, which I've been calling threads and others have called topics.  Currently the Categories and Forums are fixed, but users can post new threads and then others can add comments to those threads.

I like simple, so, using this terminology, my suggestion is to add a new category, maybe Info-ZIP Builds/Testing.  Under that, there can be forums for each major application, Zip, UnZip, and WiZ.  This adds:

* Info-ZIP Builds/Testing
  * Zip Builds/Testing      - Zip platform/OS builds.  Also post test results here.
  - UnZip Builds/Testing  - UnZip platform/OS builds.  Also post test results here.
  - WiZ Builds/Testing      - WiZ platform/OS builds.  Also post test results here.

to the above.  Looking at Zip specifically now, I was thinking that specific subjects or threads can be set up for specific things.  My current thinking is for these threads:

* Info-ZIP Builds/Testing
  - Zip Builds/Testing

    Only Info-ZIP members can create new threads in this forum.

    The below threads can be set up in each Builds/Testing forum.
    - General Information and Rules (or something like that)

      only Info-ZIP members can post here
    - New Info-ZIP Build Requests - requests for new build threads or to be added to Builders group

      anyone can post here
    - General comments/requests

       anyone can post here (this thread might get too much use)

       may have to set up additional threads for specific discussions

* Platform/port threads
  - Windows XP Platform/OS specific builds/test betas

    Info-ZIP members and designated Builders post here
    Builds can be posted here as zipped attachments
  - Windows XP Platform/OS specific builds test results

    anyone can post here
                    -- ...

With the above structure, build requests get posted by anyone, an Info-ZIP member (or forum administrator if necessary) sets up the build thread and the results thread for that port, port builders are put in the Builders group, the Builders build the port and post it to the Builds thread, and users test and comment in the Results thread.  When requested, a build can be posted on the ftp site and on SourceForge.  A build is a zip archive with the above mentioned stuff in it.  Anyway, I don't know if this all can be set up.  Comments?

**Comment by pmqs** [comment]
### Posted by E G » Tue Jul 27, 2010 3:05 am
>**Al Dunsmuir** wrote:\
>I just had another idea for distributing builds via the [ftp.info-zip.org](ftp://ftp.info-zip.org) site.
>
> The current directory structure for each platform supported by InfoZIP is:
>
> platform_name
>
>   * That directory contains the current and old formal releases
>   * Some platforms have a "beta" directory under that.
>
I would suggest that this be replaced by the following:
>
> platform_name
>
>   * release        - Contains current and old releases.
>   * beta                - Contains the tested beta release for a given official beta
>   * test                  - Contains test releases for new beta and platform code
>
The promotion path would be test->beta->release with no ambiguity.
> Al


We have almost no time to manage resources, so my preference is to do as much through the forum as possible and let users do most of the work.  However, if we are to keep things under control, there needs to be some ways for Info-ZIP to monitor and manage what is being posted and what is considered official.  At this point anything posted on the ftp site or SourceForge are considered official, and to keep things simple I'd keep it that way.  If builds are posted as attachments to the Builds threads, users can easily download them from there.

That said, the structure of the ftp site probably needs some updating.  Everything takes time, though.

**Comment by pmqs** [comment]
### Posted by E G » Tue Jul 27, 2010 3:07 am
It may make sense to have separate Builders groups for each platform.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Tue Jul 27, 2010 4:11 am
>**E G** wrote:\
>OK, I can go back and forth a little more, but I got to get back to beta development shortly.
>
>The format I'm using below is:
>
>Category
>* Forum - forum description
>  * Forum Subject (also called thread or topic)
>    * Postings in the thread (comments)
>
>Here are the existing categories available to users:
>
>* Info-ZIP Announcements
>  * Info-ZIP Releases        - Info-ZIP release announcements
>  * Site Discussions          - Problems or comments about <http://www.info-zip.org>
>
>* Info-ZIP Software Discussions and Feature Requests
>  * Info-ZIP Zip                  - Discussions and questions about Info-ZIP Zip.
>  * Info-ZIP UnZip              - Discussions about Info-ZIP UnZip.
>  * Info-ZIP WiZ                  - Discussions about Info-ZIP WiZ.0
>
>* Info-ZIP Bugs
>  * Zip Bugs                          - Please report Zip bugs here!
>  * UnZip Bugs                      - Please report UnZip bugs here.
>  * WiZ Bugs                          - Please report WiZ bugs here.
>
>Under each forum are subjects, which I've been calling threads and others have called topics.  Currently the Categories and Forums are fixed, but users can post new threads and then others can add comments to those threads.


With you so far.

>**E G** wrote:\
>  like simple, so, using this terminology, my suggestion is to add a new category, maybe Info-ZIP Builds/Testing.  Under that, there can be forums for each major application, Zip, UnZip, and WiZ.  This adds:
>
> * Info-ZIP Builds/Testing
>   * Zip Builds/Testing      - Zip platform/OS builds.  Also post test results here.
>   - UnZip Builds/Testing  - UnZip platform/OS builds.  Also post test results here.
>   - WiZ Builds/Testing      - WiZ platform/OS builds.  Also post test results here.
>
> to the above.  Looking at Zip specifically now, I was thinking that specific subjects or threads can be set up for specific things.  My current thinking is for these threads:
>
> * Info-ZIP Builds/Testing
>   - Zip Builds/Testing
>
>     Only Info-ZIP members can create new threads in this forum.
>
>     The below threads can be set up in each Builds/Testing forum.
>     - General Information and Rules (or something like that)
>
>       only Info-ZIP members can post here
>     - New Info-ZIP Build Requests - requests for new build threads or to be added to Builders group
>
>       anyone can post here
>     - General comments/requests
>
>        anyone can post here (this thread might get too much use)
>
>        may have to set up additional threads for specific discussions
>
> * Platform/port threads
>   - Windows XP Platform/OS specific builds/test betas
>
>     Info-ZIP members and designated Builders post here
>     Builds can be posted here as zipped attachments
>   - Windows XP Platform/OS specific builds test results
>
>     anyone can post here
>                     -- ...
>
> With the above structure, build requests get posted by anyone, an Info-ZIP member (or forum administrator if necessary) sets up the build thread and the results thread for that port, port builders are put in the Builders group, the Builders build the port and post it to the Builds thread, and users test and comment in the Results thread.  When requested, a build can be posted on the ftp site and on SourceForge.  A build is a zip archive with the above mentioned stuff in it.  Anyway, I don't know if this all can be set up.  Comments?



What I suggested, modified a bit to match your proposal would be:

* Info-ZIP Platform/OS Test Build Announcements
  - Info-ZIP Zip                  - Announcements about Info-ZIP Zip platform/OS test builds.
  - Info-ZIP UnZip              - Announcements about Info-ZIP UnZip platform/OS test builds.
  - Info-ZIP WiZ                  - Announcements about Info-ZIP WiZ platform/OS test builds.

* Info-ZIP Platform/OS Test Build Requests, Results and Discussions
  - Info-ZIP Zip                  - Requests, results and discussion about Info-ZIP Zip platform/OS test builds.
  - Info-ZIP UnZip              - Requests, results and discussion about Info-ZIP UnZip platform/OS test builds.
  - Info-ZIP WiZ                  - Requests, results and discussion about Info-ZIP WiZ platform/OS test builds.

I'd prefer we have a separate topic for announcing and discussing  each platform's test builds.  That's whiy I suggested mirroring the release announcements/discussion layout that folks are already familliar with.    It makes it easier for folks to find what they are looking for.

Your idea for the specific forum topics seems good - especially if there is some way that we can make the "General Information and Rules" and "New Info-ZIP Build Requests" topics "sticky" so that they always stay at the top of the list.

The platform build announcement and discussion topics shoud be a standard format, with the platform/OS name prominant.    I've found that it helps visually to put the invariant text at the front of the title and the variant text at the end, so that the variant part "pops out" visualy.  That would make it

    Test Builds: Windows XP
    Test Builds: Windows 7 and Vista 32-bit
    Test Builds: Windows 7 and Vista 64-bit

and (including the 3 fixed topics):

    General Information and Rules
    New Info-ZIP Build Requests
    General comments/requests
    Test Build Results and Discussion: Windows XP
    Test Build Results and Discussion: Windows 7 and Vista 32-bit
    Test Build Results and Discussion: Windows 7 and Vista 64-bit



>**E G** wrote:\
>When requested, a build can be posted on the ftp site and on SourceForge.  A build is a zip archive with the above mentioned stuff in it.  Anyway, I don't know if this all can be set up.  Comments?


Catch-22 warning! You can't pack away zip and especially unzip into zip executables, else how do you get them out?

Check out <ftp://ftp.info-zip.org/pub/infozip/mvs/> for a view of the current release packaging.

For the mainframe platforms, we can not use self-extracting zip files (executable format and data format are incompatible).

For the z/OS MVS port:



  * To prepare the files in the ftp directory for this port, we use the native "XMIT" utility to encode the executables, ftp them to a workstation in binary, and then ftp in binary to the infozip site.
  * Users do the opposite to instal - ftp to a workstation in binary, allocate a FB-80 format (Fixed Block, with 80 byte record length) sequential (simple) dataset with an appropriate size, and ftp into that dataset in binary, and then use the native "RECEIVE" utility to decode the executables, and receive them into the final program library.

For the VM/CMS port, the process is similar except that the native "PACK" and "UNPACK" utilities are used.

The documentation for both is packaged in ZIP files, as one normally uses that on the workstation one is using to assist with the install.

Finally, I would suggest that each test build be set up on [ftp.info-zip.org](ftp://ftp.info-zip.org) under platform_name/test/build_name with at least a core subset of the files that would be present in the platform_name/beta/beta_name and platform_name/release/release_name directories.  That way file formats and time stamps are preserved, and you can keep install instructions and build log in the same directory.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Tue Jul 27, 2010 4:13 am
>**E G** wrote:\
>It may make sense to have separate Builders groups for each platform.


Agreed.  That minimizes the possibility that someone will inadvertently use the wrong procedure to compile, package or distribute the builds.

**Comment by pmqs** [comment]
### Posted by E G » Tue Jul 27, 2010 5:23 am
>**Al Dunsmuir** wrote:\
>What I suggested, modified a bit to match your proposal would be:
>
> Info-ZIP Platform/OS Test Build Announcements
> ...
>
> Info-ZIP Platform/OS Test Build Requests, Results and Discussions
> ...
>
> I'd prefer we have a separate topic for announcing and discussing  each platform's test builds.  That's whiy I suggested mirroring the release announcements/discussion layout that folks are already familliar with.    It makes it easier for folks to find what they are looking for.


I'd prefer putting announcement threads in the Info-ZIP Releases forum.  How often do you see postings being made there?  So far there's announcements maybe monthly if users are lucky.  Even with users supplying new builds, I don't see announcements happening all that often.  In particular, I assume users looking there are expecting working code, so I would only announce reasonably working beta and release builds.

My understanding was that we are talking about builds of Info-ZIP betas and releases.  Any porting work should be coordinated with Info-ZIP so that integration issues can be minimized later.  One thing I don't want to set up is a place for unofficial feature variations to fester.  Enough of that going on off our site.  Any actual changes in how Zip or UnZip operates should be worked into the public betas so that the changes can be checked out as not breaking something else and to make sure the basic functioning of the code is consistent across ports.  If some released like object appears on the forum, users will be asking us to maintain it and we got enough to do to maintain the basic code trees.  I just think diverging the code tree is bad, except maybe to develop new features or port code that will eventually be integrated into the main trees.  Just what kind of building do you see happening and are we talking about new features being created by users?  How radical are these code changes going to be?  Who's going to integrate all this?  Or will these versions be split off?



>**Al Dunsmuir** wrote:\
>Your idea for the specific forum topics seems good - especially if there is some way that we can make the "General Information and Rules" and "New Info-ZIP Build Requests" topics "sticky" so that they always stay at the top of the list.


The first postings of a topic stay at the top.  As long as the total number of topics is limited to one page, they should always be at the top.




>**Al Dunsmuir** wrote:\
>Catch-22 warning! You can't pack away zip and especially unzip into zip executables, else how do you get them out?


True.  That's why executables are also made available separately.  At least for UnZip.



>**Al Dunsmuir** wrote:\
>For the mainframe platforms, we can not use self-extracting zip files (executable format and data format are incompatible).
>
> For the z/OS MVS port:
>
>   * To prepare the files in the ftp directory for this port, we use the native "XMIT" utility to encode the executables, ftp them to a workstation in binary, and then ftp in binary to the infozip site.
>   * Users do the opposite to instal - ftp to a workstation in binary, allocate a FB-80 format (Fixed Block, with 80 byte record length) sequential (simple) dataset with an appropriate size, and ftp into that dataset in binary, and then use the native "RECEIVE" utility to decode the executables, and receive them into the final program library.
>
For the VM/CMS port, the process is similar except that the native "PACK" and "UNPACK" utilities are used.


OK.  Can binaries be downloaded from forum attachments on these ports?  Seems if the above rules are followed, it might work.  I'd rather only put tested stuff on the ftp site.



>**Al Dunsmuir** wrote:\
>Finally, I would suggest that each test build be set up on [ftp.info-zip.org](ftp://ftp.info-zip.org) under platform_name/test/build_name with at least a core subset of the files that would be present in the platform_name/beta/beta_name and platform_name/release/release_name directories.  That way file formats and time stamps are preserved, and you can keep install instructions and build log in the same directory.


I think tested betas and releases are all that should be put on the ftp site.  Anything rougher than that may be buggy and should not be advertised to unknowing users as something they want to download.  What do you expect general users to do with rough test builds?  And why can't actual developers just use attachments on the forum to distribute them?  You're not making your point sufficiently.  I see one of the differences between rough builds and public betas are the public betas are tested for user use.  What you're calling a build is essentially what we already maintain as internal betas.  Are you suggesting that a maintainer like me should maintain both internal betas and forum builds?  Sounds like you're trying to set up a way to develop independent versions of Zip and UnZip.  It just looks like we're setting up for a mess in a few months trying to integrate all this together, which may do more harm than good.  Unless these builds will be split off as non-Info-ZIP projects later on, in which case why are we supporting them.  Can you in simple language describe just what the end goals of the builds are and give examples of how they will be used on a day-to-day basis?  Unless it all comes together eventually as a single set of products that support users across all ports, it seems we're going in the wrong direction.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Tue Jul 27, 2010 8:30 am
Sleep is for the weak.

>**E G** wrote:\
>I'd prefer putting announcement threads in the Info-ZIP Releases forum.  How often do you see postings being made there?  So far there's announcements maybe monthly if users are lucky.  Even with users supplying new builds, I don't see announcements happening all that often.  In particular, I assume users looking there are expecting working code, so I would only announce reasonably working beta and release builds.


I'd originally thought you wanted to keep the announcements for zip and unzip beta and final releases in a separate category from the platform/OS builds.  If that is no longer the case, then it does indeed make sense to just have just the current announcements Category, and add the Platform/OS build discussion and request section.

At absolute most frequent, I can see weekly updates to the z/OS MVS, z/OS USS and VM/CMS ports while the platform specific code is under heavy development.  Once the core support for the various file types is complete, it would slow down significantly to match the pace of changes in the zip/unzip core codebase (beta releases) and new platform-specific function.

If we can get some platform build folks involved, I can see things getting quite noisy around beta release events, and around formal releases.  Very cyclical.

If I got the count right, there are 23 platforms on the ftp site.  Some are likely dead, but others are likely missing.  Within each plaform, there are multiple flavours with identical code that may be built by different folks within a particular build team eg Windows (XP, 7/V 32-bit, 7/V 64-bit, CE), Linux (i686, X86_64, ARMv5, etc).  Call it 30 different combinations of platform/OS, with one topic each.  At 15 topics per page, it still would be only two pages.


>**E G** wrote:\
>My understanding was that we are talking about builds of Info-ZIP betas and releases.  Any porting work should be coordinated with Info-ZIP so that integration issues can be minimized later.


For the z/OS MVS, z/OS USS and VM/CMS ports, the additional function is purely there to make the core code usable on the platform.  This means supporting the invocation environment (including the proposed mapping of the parser to UTF-8, various native POSIX and non-POSIX filesystems, encode and extract file name mapping, and finally archive file data character translation.

It will take a lot of code to support the functions that RBC and other users such as lutz and fits have identified, but once that it done it should not have to change very much over time.    Making the name mapping and data character translation code as generic as possible means that other platforms with similar requirements can have their problems solved at the same time.  Then again, our platforms may represent "the problem children".



>**E G** wrote:\
>One thing I don't want to set up is a place for unofficial feature variations to fester.  Enough of that going on off our site.


Aye, and the GnuWin32 zip and unzip packaging is a perfect example.  Like the other GnuWin32 programs, it uses a freeware installer that allows the display of a logo, acceptance of a licence (in this case, the InfoZIP licence), the optional download of the relevant source code, and finally the install of the zip program and docs.  It even supports uninstalls.  Very smooth and polished.

Unfortunately, whoever did it did not keep up with the new zip/unzip releases. If the new builders for the Windows platforms use that install as the base (source and setup was available), the InfoZIP project will suddently have something spiffy to show off.

I expect folks on OS/X want standard installs. and I know for sure that the Linux folks want standard rpm and deb packaging.



>**E G** wrote:\
>Any actual changes in how Zip or UnZip operates should be worked into the public betas so that the changes can be checked out as not
> breaking something else and to make sure the basic functioning of the code is consistent across ports.


There needs to be a clear separation between core zip/unzip functionality and platform-specific code that brings zip/unzip to a given platform.  Platform-specific code needs to be well documented and clearly written to allow folks familliar with that platform to take over a given port when required.

I don't expect your group to maintain (or understand below a very generic level) the z/OS and z/VM port platform specific code, no more than I would expect you to maintain Tandem or iSeries platform-specific code.  It will take a lot of new platform code to allow the z/OS and z/VM ports to fully support the rather wierd externals (files and character translation) for these platforms.    The salient point is that once this is all in place, the focus returns to the core zip/unzip code.

At that point, I can back off my contribution to more of a platform builder role... or help on other ports on my own time.



>**E G** wrote:\
>If some released like object appears on the forum, users will be asking us to maintain it and we got enough to do to maintain the basic code trees.


I believe you need to have platform builders (and to some extent, developers) in place to maintain that side of zip and unzip.   Ports are either maintained or unmaintained... and the latter is really the problem of the users on a platform to solve.  TANSTAAFL.
Your focus has to stay on the common command logic and core zip and unzip code.



>**E G** wrote:\
>I just think diverging the code tree is bad, except maybe to develop new features or port code that will eventually be integrated into the main trees.  Just what kind of building do you see happening and are we talking about new features being created by users?  How radical are these code changes going to be?  Who's going to integrate all this?  Or will these versions be split off?


Our needs at RBC are entirely for platform-specific code to fully support the filesystems we use, support the character sets (and code pages) we use, and a few functions (tab expansion, tab encoding, blank padding and trailing blank trimming) that help support users on our platform in common tasks where there is no shell available to do the equivalent data transformations.

zip and unzip are just one aspect of the task of managing data transfer and transformation within our area of our business.  As I said, my primary job is heavily tied to solving these problems in general on z/OS.



>**E G** wrote:\
>The first postings of a topic stay at the top.  As long as the total number of topics is limited to one page, they should always be at the top.


It looks like if all platforms have active builds we will eventually grow to be more like 2 pages.  Forum software often supports "sticky" topics that stay at the top... and that is why i suggested having Hunter look into it.  No immediate rush.



>**E G** wrote:\
>True.  That's why executables are also made available separately.  At least for UnZip.
>
> OK.  Can binaries be downloaded from forum attachments on these ports?  Seems if the above rules are followed, it might work.  I'd rather only put tested stuff on the ftp site.


While I prefer ftp in all cases, I can see your need to differentiate platform (unit or system verification) test builds from fully tested and polished builds representing full support for a beta or release level for that platform.


>**E G** wrote:\
>I think tested betas and releases are all that should be put on the ftp site.  Anything rougher than that may be buggy and should not be advertised to unknowing users as something they want to download.


I would hope that builds undergo at least some minimal unit tests before being posted, but if the builder does not have the environment to support testing a particular item (EG. me and BFS files with the VM/CMS port) they need to use the forum to work with the platform testers.



>**E G** wrote:\
>What do you expect general users to do with rough test builds?  And why can't actual developers just use attachments on the forum to
> distribute them?  You're not making your point sufficiently.


I guess my preference for FTP and dislike for web page attachments isn't enough, eh?  Your point is made.



>**E G** wrote:\
>I see one of the differences between rough builds and public betas are the public betas are tested for user use.  What you're calling a build is essentially what we already maintain as internal betas.


The problem is that there are currently very few builds of any sort available from [ftp.info-zip.org](ftp://ftp.info-zip.org).
I've seen comments in the Fedora development mailing lists wondering if the zip and unzip upstream is still active.    Unless those internal builds become available for folks to test, they don't really exist for the users.    That's what can lead to forking and distribution-specific ad-hoc solutions not in the interest of the zip and unzip developers or user community.



>**E G** wrote:\
>Are you suggesting that a maintainer like me should maintain both internal betas and forum builds?


It's entirely up to you if you want to maintain a port in addition to working on the core zip code.   I'd suggest that the latter is too important to take that time away working on the former, but it's your time and committments.



>**E G** wrote:\
>Sounds like you're trying to set up a way to develop independent versions of Zip and UnZip.


No.  Rather the opposite.

The same threads in the Fedora developer forum also touched on Fedora efforts to follow their build policy and eliminate imbedded versions of libraries such as zlib.  They had been frustrated that projects seemed to be imbedding zlib variants with relatively small changes that were not accepted upstream.    Perhaps  the projects had not worked very effectively with the upsteam team.  Or perhaps they were frustrated with the very long release cycles.

I know you folks are not the zlib maintainers (having found [www.zlib.net](http://www.zlib.net)), but the issues the Fedora folks are still present for zip and unzip.  Making it easier to do more frequent releases of zip and unzip (with smaller amounts of incremental chagnes) would be very helpful in making sure that everyone considers the project alive and well and helps give time for the core team to consider useful suggestions.

>  It just looks like we're setting up for a mess in a few months trying to integrate all this together, which may do more harm than good.  Unless these builds will be split off as non-Info-ZIP projects later on, in which case why are we supporting them.  Can you in simple language describe just what the end goals of the builds are and give examples of how they will be used on a day-to-day basis?  Unless it all comes together eventually as a single set of products that support users across all ports, it seems we're going in the wrong direction.

Excuse me? Where the heck is this coming from?

Where have I said anything that is not related to getting useful z/OS MVS, z/OS USS and now VM/CMS ports of zip and unzip running?  The old ports are barely functional (last VM/CMS one in 1998, and there is NO current z/OS MVS port that works other than our internal RBC one (USS-specific changes broke the original MVS port). Working sideways via z/OS USS to get minimal broken function is not accptable.  It is essentially unchanged from when I first approached the forum in 2008, but in fact growing worse now due to new breakage.
I volunteered to take over the VM/CMS port because of all the code it shares (or could share, in some case) with the z/OS ports and the need to regression test the effect of these changes.

As I posted months ago, we need the z/OS ports to be fully functional, solid and perform well, and provide ZIP64 support.  We need them to be widely available (via the InfoZIP web site) so our partners can download and install the port without us personally supplying it to them.  We also need these ports to be compatible with the other zip/unzip ports for partners who supply or consume data on other platforms such as Windows, Linux and AIX.

Finally, as I have also said recently, I'm also interested in helping out with the AIX and Windows CE ports.

Anything reasonable that we can do to help ensure that zip and unzip are readily available, solid and actively developed just makes sense.

**Comment by pmqs** [comment]
### Posted by E G » Wed Jul 28, 2010 1:43 am
Al, as we later discussed off line, I'm OK with the general approach being taken, I just was trying to clear up in my head what you were saying.

We do maintain nearly all the ports ourselves.  The only real exceptions are the old Mac port, which is no longer being updated by the original author and has been replaced by the Unix port for Mac OS X, and the MVS/CMS/zOS port where our group expert lost access to a mainframe years ago and then left the group.  We try, in fact, to be responsible for all code we post.  (Though note all code we post is subject to the Info-ZIP license.)  We also try to address questions from users regarding the functioning of our applications on any port we support and fix bugs on any port.  Hopefully now you can provide some actually useful responses for us when users ask about the MVS and z/OS ports.

With regards to the apparent lack of responsiveness of the group, we had internal issues mainly centered around lack of support of the UnZip Maintainer.  He has apparently dropped off the radar now and I've had to pick up the slack as of early this year.  Well, actually things are moving much faster now, compared to not at all before.  But I'm really stretched.  I'd appreciate being cut some slack as I try to get both Zip and UnZip back on a continuous development schedule while also performing my other group duties.  Also note that we had some ftp and web server outages, needed to move servers more than once, and had other issues pop up in the last few years.  These all took up time.  There is quite a lot of history regarding this group you don't know yet.  Perhaps we should discuss this more off line.

Sorry if anything I said was taken offensively or was disruptive.  I was quite tired at the time.  Let's both take a deep breadth and focus on the reason we are here, to get this code developed, working, and out.

**Comment by pmqs** [comment]
### Posted by Al Dunsmuir » Wed Jul 28, 2010 3:50 am
I think we had a case of two very tired people attempting to interact  (yesterday was a 25 hour day for me).  We really didn't understand each other's style, or motivations.  I do think it has helped us understand each other a bit better.

Stuff happens - spilled milk under the bridge as one manager at IBM used to say.

I'm not low key, or subtle.  I know it - one gent with whom I had a week to do a brain-dump style product transfer said that interacting with me in person was "like trying to drink from a firehose".   Best bet in interacting is to simply indicate when I need to pause so you can digest what I've said.    That also tends to work well in mailing lists when one can not throw up the STOP sign.  I can move on to working offline, or posting on a different topic.

My intent is to get up to speed as quickly as possible (after being offline from the forum for nearly two years).  I spent about two months doing so slowly, but for the last two weeks I've been spending a lot of time posting and trying to keep up with the various responses, plus cover all previous discussion/issues with MVS/USS.    You are in turn were trying to keep up with me doing that, and you see yourself falling behind.  Oops... disaster waiting to happen.

I think it would be an unreasonable expectation on the developer (and perhaps dangerous in outcome) to have folks not working with z/OS on a daily basis working on the z/OS (especially MVS) platform code.  It is however entirely reasonable for your team of experts in the core zip and unzip codebase and other platforms to be able to at least navigate the platform code and gain a reasonable understading on what is being done (at a high level) and why.   This allows a security audit where the core team can validate the platform code in a reasonable way with the assistance of folks familliar with the platform.

One of the things that I suggested I need to initiate as part of this process is a "InfoZIP ZIP/UNZIP Platform Guide" with chapters for each port.  I can do my ports, and make a start on the platforms with which I am familliar.  Then the developer responsible for those ports can tidy/complete this information and add information for their other ports.

I should make sure that my efforts in the forums and mailing list are well focussed.

I should also learn to pace myself on the forums, and work on the other things on my plate


* I need to do my other work.

  The good news is that both that and zip/unzip are big projects and my delivery targets are reasonably flexible for the former.

* I need to dedicate some time to
play with my fun projects so I can unwind and not come across as a total
caffine junkie :smile:.

* I need to give you the information you mentioned you need.

  I've tried to provide an off-the-cuff overview that hopefully provides some insight, but may confuse further.



* I need to get up to speed with things like the current zip/unzip beta codebase.


* I also need to learn how to managing changes with diff (so can generate a proper tab-encoded diff, vs the tab expanded source I now have).


* Finally, I need to set up a local git instance on my home development server (which I also need for other projects) to manage the platform-dependent changes until things all get integrated and to assist in my development work.

  git has useful and powerful tools for multiway change management.    It should help me migrate my changes originally against the old source base into useful changes against the current source base.  I can then restructure those changes to fit the new demands of the z/OS and VM/CMS ports.  It should also allow me to rebase my platform code against new beta levels as they come out.

**Comment by pmqs** [comment]
### Posted by E G » Thu Jul 29, 2010 1:54 am
My thought is well commented and documented code should be understandable by any coder.  If special OS stuff is done, that should be well described in the code, what is being done and why, and the possible effects.  However complex your stuff gets, if you can't describe it in terms that a non-MVS coder can understand, then that may be time to pause and think of a simpler approach, or get help to describe what's going on.  So in that sense, we all shouldn't have to be experts in MVS coding to review and maybe debug the MVS parts of the code.  That should be a goal anyway.  I believe that, if you can't describe it, you probably didn't fully understand it in the first place, and that is not good code to put into an application.



 ## Post by Al Dunsmuir » Thu Jul 29, 2010 2:09 pm

I was taught (and have learned the hard way over time) that well structured code flows from a process like this:

1) Determine at a high level what you want to do, and attempt to express the task clearly, first at a high level and then working down to a lower level.



   * If you can't do so, there is something missing that needs to be clarified.
   * Sometimes the details can be worked out later, but sometimes the detail is the crux of the matter.
   * It is also important to determine how processing is to be driven - strictly procedural, driven by events (external, or within the input).  For event-driven programs, it is useful to use a state-machine approach during processing.

2) Determine how processing is going to be broken up.



   * Some problems are best solved by long linear processing.
   * Others need to be broken up into smaller routines to be able to clearly understand the data processing flow.
   * Try to recognize patterns that can be solved by common routines

3) Identify the differences between externally visible and internal processing and data.



   * If you need to expose internal data, do so carefully and consistently with a mapping/transformation that allows the internal processing to continue to evolve over time.

3) Come up with descriptive consistent names for the functions

   * This helps you understand the code when reading it.

4) Functions should have a standard descriptive comment block that describes what they are, what they do, the input and output parameters.


   * Trivial functions may have very little documentation beyond that
   * Complex functions need to have some notes after the initial block that describes the internal logic.

5) When coding on a given function, use useful comments.



   * I often put in all the comment blocks first, and then fill in the code around them.
If you can't describe the logic clearly in your comment blocks, you are not ready to code.

6) Choose variable names that are consistent and appropriately descriptive

   * Use simple names for things like loop indicies.

7) Code defensively



   * Functions are often passed invalid parameters
   * Functions also are often used outside their original design, as the project evolves and adds function.  Do reasonable things when presented with input, and think about recovery.

As code evolves, restructure/refactor/rewrite as required.

* Patchwork spaghetti code often works as bad as it looks.
* If you rewrite it clearly reflecting the new reality you are ready for the next reality change

I tend to code (and document) so that the next person that comes along can pick up the code and work with it relatively easily.    As I've worked on a number of projects that have grown beyone 250 KLOC and spanned 5 to 10 years, I have often been "the next person" a few years after I last touched a given area.


 ## Post by Al Dunsmuir » Thu Jul 29, 2010 2:43 pm

Ed,

I've been thinking about one aspect of how zip and unzip on both z/OS and VM/CMS have two modes of operation.


* Non-POSIX

  When executed from the non-POSIX command line, the
default conventions for file names is that of the native non-POSIX file
names.

  POSIX files can be identified by context (DD:Ddname that points to a USS file), or by a prefix (leading single slash "/", ".". or "~")

* POSIX

  When executed from the POSIX command line, the default conventions for file names is that of the USS POSIX file names.

  Non-POSIX files can be identified by a prefix (leading double slash ("//") or by a DD:Ddname that points to a non-POSIX file).

My original thought for the z/OS side was for there to be two distinct executable programs, with a hard-coded default mode of operation for each.

It is simpler to instead use the mechanism where the USS "program" is actually what is called an "external link". This link points to the non-POSIX executable program.   During execution via either link, the program logic can check whether arg[0] shows it was invoked using the POSIX file convention, and mark the required default file type.

We should be able to use this approach on both z/OS and VM/CMS, so that there is only one port for a given platform.  There would be a 1-time installation step to set up the external link in a POSIX directory within the executable search path.  There will be a lot of shared code between the two platforms.

Each of these two ports would support all the types of files for that platform - z/OS (MVS non-POSIX, USS POSIX) and VM/CMS (CMS non-POSIX, USS POSIX, SFS non-POSIX) respectively.

Each file would be tagged with it's type.  Once things are finalised, as you noted this has to be documented in the application notes document.

This should simplify the task ahead a bit.

Al

 ## Post by E G » Mon Aug 09, 2010 3:54 am

Perhaps.  I find it's usually hard to say what will make the work easier until you've already done much of the work.

By the way, when developing Zip 3.0 there was a point where we realized we were going down a dead end (the code was getting way too complicated) and regressed the code back a couple betas and started over.  That was a few months of work we just threw out, but it was a good decision in the long run.  So as long as we're willing to start over if needed and scrap a bad approach once we see it's bad, we should get to the end of the tunnel.

**Comment by pmqs** [comment]
>[!NOTE]
The text in this issue was sourced from [here](https://web.archive.org/web/20110222210215/http://www.info-zip.org/phpBB3/viewtopic.php?f=3&t=323)
